### PR TITLE
bpf2c: emit maps in order they appear in ELF file

### LINKED ISSUE #875: bpf2c should emit maps in the order they are present in the ELF file.
### Describe the bug

Using `std::map` causes the map entries to be in different order in the map than they were in the ELF file. Some map-in-map implementations use `inner_map_idx` to identify the inner map which is the index of the map in the maps section. If maps get re-ordered while storing in `map_definitions`, the `inner_map_idx` will become invalid.

### Affected OS

- [ ] Windows Server 2022
- [ ] Windows Server 2019
- [ ] Windows 11
- [ ] Windows 10
- [ ] Windows Insider Preview (specify affected build below)
- [ ] Other (specify below)

### Additional OS information

_No response_

### Steps taken to reproduce bug

NA

### Expected behavior

The maps should be stored in the order as they appear in the ELF file.

### Actual outcome

The maps stored in `map_descriptors` can be re-ordered.

### Additional details

_No response_

### PR DESCRIPTION
**Pull request by saxena-anurag**
## Description

This PR does the following:
1. Save the original map index while parsing the maps section.
2. Print the maps in the same order when emitting the C code.

## Testing

Added new test in api_test.

## Documentation

NA

Fixes #875

### CODE REVIEW

File: tests/api_test/api_test.cpp [hunk 1]
```diff
@@ -562,6 +562,17 @@ TEST_CASE("bindmonitor_tailcall_native_test", "[native_tests]")
...
+    if (outer_idx_map == nullptr)
+        cleanup();
+    REQUIRE(outer_idx_map != nullptr);
+
+    int outer_idx_map_fd = bpf_map__fd(outer_idx_map);
+    if (outer_idx_map_fd <= 0)
+        cleanup();
+    REQUIRE(outer_idx_map_fd > 0);
+
     // Cleanup tail calls.
```
Comment (dthaler) [inline tests/api_test/api_test.cpp:576]:
```suggestion
    // Clean up tail calls.
```

File: tools/bpf2c/bpf_code_generator.cpp [hunk 2]
```diff
@@ -216,6 +216,14 @@ bpf_code_generator::parse()
     ELFIO::const_symbol_section_accessor symbols{reader, reader.sections[".symtab"]};
 
     if (map_section) {
+        size_t data_size = map_section->get_size();
+
+        if (data_size % sizeof(ebpf_map_definition_in_file_t) != 0) {
```
Comment (dthaler) [inline tools/bpf2c/bpf_code_generator.cpp:221]: In [verifier PR 295](https://github.com/vbpf/ebpf-verifier/pull/295) I put a more intelligent algorithm for determining the map definition size in the file in https://github.com/vbpf/ebpf-verifier/blob/master/src/asm_files.cpp#L79 which is the algorithm already used by libbpf.

Consider using that algorithm, which supports older (smaller) structures by 0-padding the rest of the fields.

**Review by dthaler — DISMISSED** [review]

**Review by dthaler — APPROVED** [review]

**Comment by Alan-Jowett** [comment]
Can we modify tests\bpf2c_test_wrapper to have it generate code that uses map-in-map? Currently, we don't have coverage for the map-in-map code generation case.

**Comment by saxena-anurag** [comment]
> Can we modify tests\bpf2c_test_wrapper to have it generate code that uses map-in-map? Currently, we don't have coverage for the map-in-map code generation case.

I can do that. I already have added map-in-map test cases in api_test.c though (`bindmonitor_tailcall_native_test`). Also, once the changes for #852 are in, we should be able to run all the existing unit tests for native mode also.

**Comment by Alan-Jowett** [comment]
> > Can we modify tests\bpf2c_test_wrapper to have it generate code that uses map-in-map? Currently, we don't have coverage for the map-in-map code generation case.
> 
> I can do that. I already have added map-in-map test cases in api_test.c though (`bindmonitor_tailcall_native_test`). Also, once the changes for #852 are in, we should be able to run all the existing unit tests for native mode also.

OK, as long as it gets covered, I am not picky about how we do it.

File: tools/bpf2c/bpf_code_generator.cpp [hunk 3]
```diff
@@ -241,15 +249,11 @@ bpf_code_generator::parse()
                 }
                 map_definitions[symbol_name].definition =
                     *reinterpret_cast<const ebpf_map_definition_in_file_t*>(map_section->get_data() + symbol_value);
+
+                map_definitions[symbol_name].index = symbol_value / sizeof(ebpf_map_definition_in_file_t);
```
Comment (shankarseal) [inline tools/bpf2c/bpf_code_generator.cpp:253]: for ease of understanding can this be renamed to `map_section_offset` ?

**Review by shankarseal — APPROVED** [review]

**Pull request merged**